### ISSUE TITLE
Add query max-memory-per-node usage to query completion event 

### ISSUE THREAD

**Issue by OleksiiDuzhyi** [prestosql/presto#1109]
Hi,

We use Presto extensively for ETL.
We are collecting query stats from QueryCompletedEvent and we'd like to have an ability to identify queries that are close to `query.max-memory-per-node` configuration in order to redesign/rework them and avoid failures of critical ETL during night hours. 

Thank you.

**Comment by sopel39** [comment]
It would be definitely a valuable addition. We have all memory information in stats, but we would need to aggregate that by node to compute peak.

**Comment by Terry93** [comment]
Hello, I would like to help with working on this. Thanks!

**Comment by findepi** [comment]
@Terry93 appreciated, go for it!

**Comment by Terry93** [comment]
@sopel39 @findepi Hello I tried to find out the potential solution for this issue but it seems not that straightforward to me, may I have some inputs from you?

Take getting the peak user memory for a query on a worker node for example (system memory should be the same), I think the best way to get this information using the `queryMemoryContext` (getUserMemory function) in the `QueryContext.java`. We can somehow keep updating the peak user memory.

However, I am not sure what is the best way to pass this info back to the Coordinator (since we want to deliver this info to Coordinator, either to UI / QueryCompleteEvent). Since this sounds like query level information on a worker, thus maybe we need to expose it on workers `/v1/query` endpoint (which is something does not exist).

I am relatively new to Presto dev, so your suggestions will be helpful! Thanks!

**Comment by Terry93** [comment]
@sopel39 @findepi Gentle ping, can I have some opinions from you guys? Thanks!

**Comment by sopel39** [comment]
> Take getting the peak user memory for a query on a worker node for example (system memory should be the same), I think the best way to get this information using the queryMemoryContext (getUserMemory function) in the QueryContext.java. We can somehow keep updating the peak user memory.

We could add a `NodeStats` class (populated by `QueryContext`) and send that back to coordinator. Then we could add `List<NodeStats>` as part of `QueryStats` class.

**Comment by Terry93** [comment]
> > Take getting the peak user memory for a query on a worker node for example (system memory should be the same), I think the best way to get this information using the queryMemoryContext (getUserMemory function) in the QueryContext.java. We can somehow keep updating the peak user memory.
> 
> We could add a `NodeStats` class (populated by `QueryContext`) and send that back to coordinator. Then we could add `List<NodeStats>` as part of `QueryStats` class.

Thanks! I will try this way!

**Comment by Terry93** [comment]
@sopel39 Hello can you share more on how the worker sends the NodeStats back to the coordinator? Is there any existing skeleton already built up for this needs or an example I can refer to?

**Comment by sopel39** [comment]
@Terry93 Right. It seems that there is missing endpoint for exposing node query stats.
We currently reconstruct query stats from stage and task stats, but these cannot be used to compute node query stats due to it's asynchonrious manner.
@dain any suggestions here?
I think we would need to add a new endpoint for retrieving `NodeQueryStats` which will be used by `QueryStateMachine` to keep node query stats.

**Comment by electrum** [comment]
For tracking, peak usage, I'd add `peak` fields to `QueryContext` and have `updateUserMemory()` and `updateSystemMemory()` update the peak fields.

There's not really a concept of a "node" for a query. The coordinator only sends tasks (single stages) to workers. I think it makes sense to return the peak memory as part of the task information, probably in `TaskStats`. The coordinator would then choose the largest value (doesn't matter which node it comes from).

**Comment by sopel39** [comment]
> There's not really a concept of a "node" for a query. 

@electrum Although this is true having node query peak memory utilization would be useful to identify queries close to memory limit per this issue descrition

**Comment by electrum** [comment]
@sopel39 I agree it’s useful. My point is that we already have infrastructure to track task information, so it seems better to utilize that, rather than creating a new channel. If we later need more query/node information, it might make sense to create a new channel and move this there. But that seems overkill for this purpose.

@dain might have more thoughts here.
